lib/core/src/Cardano/Wallet/DaedalusIPC.hs

```diff
@@ -0,0 +1,231 @@
+    deriving (Show, Eq, Generic)
+
+aesonOpts :: Options
+aesonOpts = defaultOptions { sumEncoding = ObjectWithSingleField }
+
+instance FromJSON MsgIn where
+    parseJSON = genericParseJSON aesonOpts
+instance ToJSON MsgOut where
+    toEncoding = genericToEncoding aesonOpts
+
+daedalusIPC :: Int -> IO ()
+daedalusIPC port = withNodeChannel (pure . msg) action >>= \case
+    Right act -> do
+        sayErr "[INFO] Daedalus IPC server starting"
+        act
+    Left NodeChannelDisabled -> do
+        sayErr "[INFO] Daedalus IPC is not enabled"
+        sleep
+    Left (NodeChannelBadFD err) ->
+        sayErr $ "[ERROR] Starting Daedalus IPC: " <> err
+  where
+    -- How to respond to an incoming message, or when there is an incoming
+    -- message that couldn't be parsed.
+    msg (Right QueryPort) = Just (ReplyPort port)
+    msg (Left e) = Just (ParseError e)
+
+    -- What to do in context of withNodeChannel
+    action :: (MsgOut -> IO ()) -> IO ()
+    action send = send Started >> sleep
+
+    sleep = threadDelay maxBound
+
+----------------------------------------------------------------------------
+-- NodeJS child_process IPC protocol
+-- https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options
+
+data NodeChannelError
+    = NodeChannelDisabled
+      -- ^ This process has not been started as a nodejs @'ipc'@ child_process.
+    | NodeChannelBadFD Text
+      -- ^ The @NODE_CHANNEL_FD@ environment variable has an incorrect value.
+    deriving (Show, Eq)
+
+withNodeChannel
+    :: (FromJSON msgin, ToJSON msgout)
+    => (Either Text msgin -> IO (Maybe msgout))
+       -- ^ Incoming message handler
+    -> ((msgout -> IO ()) -> IO a)
+       -- ^ Action to run
+    -> IO (Either NodeChannelError (IO ()))
+withNodeChannel onMsg handleMsg = fmap setup <$> lookupNodeChannel
+  where
+    setup handle = do
+        chan <- newEmptyMVar
+        let ipc = ipcListener handle onMsg chan
+            action' = handleMsg (putMVar chan)
+        race_ action' ipc
+
+-- | Parse the NODE_CHANNEL_FD variable, if it's set, and convert to a
+-- 'System.IO.Handle'.
+lookupNodeChannel :: IO (Either NodeChannelError Handle)
+lookupNodeChannel = (fromMaybe "" <$> lookupEnv "NODE_CHANNEL_FD") >>= \case
+    "" -> pure (Left NodeChannelDisabled)
+    var -> case readEither var of
+        Left err -> pure . Left . NodeChannelBadFD $
+           "unable to parse NODE_CHANNEL_FD: " <> T.pack err
+        Right fd -> tryJust handleBadFd (fdToHandle fd)
+  where
+    handleBadFd :: IOException -> Maybe NodeChannelError
+    handleBadFd = Just . NodeChannelBadFD . T.pack . show
+
+ipcListener
+    :: forall msgin msgout. (FromJSON msgin, ToJSON msgout)
+    => Handle
+    -> (Either Text msgin -> IO (Maybe msgout))
+    -> MVar msgout
+    -> IO ()
+ipcListener handle onMsg chan = do
+    hSetNewlineMode handle noNewlineTranslation
+    catch (race_ replyLoop sendLoop) onIOError
```

Comment:
Nice pattern. I'd have a slight preference for `concurrently` instead of `race_` since we aren't actually expecting any of the two sides to eventually terminate (which `race_` conveys). We still get the cancellation behavior with `concurrently` if I am correct.

Reply:
It really doesn't matter which one. Both of them imply the wrong thing. But if `concurrently` makes you happy, then I'll change it.